### ISSUE TITLE
Rerun dedup capbilities

### ISSUE THREAD

**Issue by dbekaert** [aria-tools/ARIA-tools#198]

**Comment by sssangha** [comment]
I will outline my motivation and proposed strategy behind adding in the ability to re-run the codes efficiently without restarting from scratch everytime. In the end, we should only start regenerating deleting all existing outputs from scratch when the user changes input argument (s) (i.e. different bounding box, different mask, different multilooking factor, etc.).

For each new ARIA-tools call, I propose writing a sorted dictionary of user input arguments to file (omitting the reference to the number of threads and verbose mode, which aren't relevant).
Behind this reference record, I have brainstormed a check and filter in the main ```ARIA_standardproduct``` class that is setup after the spatiotemporal sorting subfunction and behaves as outlined below for whenever a user relaunches an ARIA-tools call:
1. Compare new user inputs with the aforementioned record of user input arguments stored from the previous run. 
    - If they are the same then do a walk of the output directory to get a list of products already produced, compare it with the dictionary of IFGs passed by the class and purge references to existing products. Thus, only new IFG products are to be extracted. 
    - If certain arguments are different (e.g., mask, multilooking, etc) then regenerate all outputs from scratch.
    - However, if some specific arguments are different (e.g. specified layers, tropospheric corrections) then just generate new requested products. E.g. if a user wishes to now extract amplitude layers (whereas before only coherence was specified) then just extract the amplitude layers without regenerating coherence outputs. If tropospheric corrections were requested now and weren't before, just generate those corrections without regenerating anything else.
2. If no differences are flagged, then just print a message stating input arguments haven't changed, thus outputs will not be modified.

I envision that this routine would be a separate subfunction of the ```ARIA_standardproduct``` class that is only triggered when toggled by a separate argument in the main ```ariaExtract``` or ```ariaTSsetup``` calls, since it obviously wouldn't make sense to filter out products for ```ariaPlot``` calls which generate figures illustrating cumulative metrics.

This is the cleanest logic I could formulate at the moment, and indeed perhaps it would make more sense once I have a working version implementing such proposed changes. But if any of you have specific comments and requested features/adjustments of relevance to these efforts, please discuss them here.

**Comment by sssangha** [comment]
Storing a physical dictionary of inputs arguments would not require much overhead. However a clean, yet less transparent alternative would be to generate a unique hash and store it to a hash file that may be referenced vis-a-vis successive user inputs.

Just as a note, from what I've seen it is recommended that such a dictionary of input arguments is to be serialized as JSON before hashing. E.g.:
```
import hashlib
import json

a={'intputfile':'/path/to/file', 'bbox':'-9 9 10 12'}
print(hashlib.sha1(json.dumps(a, sort_keys=True).encode()).hexdigest())
```
Source: https://stackoverflow.com/questions/16092594/how-to-create-a-unique-key-for-a-dictionary-in-python

Any strong feelings one way or the other? I.e. for storing hash files as opposed to a full dictionary of input arguments (which should be very small anyways). I personally prefer writing out a full dictionary as I outlined above because it is more transparent, but am open to discussion.

**Comment by dbekaert** [comment]
# Variables of relevance for evaluating a change for reprocessing 

@sssangha Lets also dig up the discussion on ARIA-tools slack channel and make a table per layer on what are parameters that can change and which cannot. You can edit my message here:


| Layer | if these variable changed needs updating | additional comments |
| ------------- | ------------- | ------------- |
| unwrapped phase/connected component | stitch-method, cut line, mask, multi-look  |   |
| coherence/amplitude | mask, cut-line, multi-look |  |
| 3D cube variables | mask, cutline, DEM input, multi-look |  |
| DEM | cutline, DEM input, multi-look |  |
| water mask | cutline,  multi-look | |
| mask | user-provided mask, water mask, amplitude threshold, multi-look|  |



The cutline is driven by different variables:
- bbox (SNWE, shapefile)
- union/intersection
- mo


I believe that the cutline could be translated as part of the evaluation. Its not as transparent as it depends on many different variables. But there are scenarios where a user could take a certain bbox and then forced it with mo to still have the original cutline. Under that scenario the arguments have changed but the cutline did not. Same with respect to the next item below:



# Option to continue processing
Additional to the ability to dedub, we should include capabilities that allows for easy expansion of existing stack without the need for the user to set-up the input arguments. e.g. perhaps --expand option which would then load the existing cutline, calculate a -mo value and forces 100% overlap or so for that cutline. This does  only capture if a user want the expansion to be 100% covered, but would is a good point to start the discussion. 

See here for how `-mo` could be calculated: https://github.com/aria-tools/ARIA-tools-docs/blob/aria_sds/JupyterDocs/ariaEnd2End/ariaEnd2End.ipynb


# storing of processing variables
In terms of storing the input arguments that should be done in line with the issue ticket on logging #94. 
@Askaholic Could we make a specific logging channel to a file which woudl capture all the inputs a user has per program with teh time the program was called?

**Comment by dbekaert** [comment]
I believe that the cutline could be translated as part of the hash. its not as transparent as it depends on many different variables. But there are scenarios where a user could take a certain bbox and then forced it with mo to still have the original cutline. Under that scenario the arguments have changed but the cutline did not.

**Comment by Askaholic** [comment]
I'm not sure why you would need to use logging for that. I would just write to a json file using dump:
```python3
if __name__ == "__main__":
    parser = create_parser()
    args = parser.parse_args()

    if args.modify_last_run:
        with open("user-inputs.json") as f:
            # Restore args to the previous state
            args.__dict__.update(json.load(f))
    else:
        with open("user-inputs.json", "w") as f:
            json.dump(f, args.__dict__)

    # Call main routine...
```
Note that this will only work if your args consist entirely of json serializable values (str, int, bool, float, None, list, dict). If you are doing fancier arg parsing that may store arbitrary python objects, then you can use pickle instead (although be aware than this has some security implications).

As for the discussion about using a hash, here are the main desirable properties of hashes:
1. They are irreversible
2. They are deterministic
3. They are small

I'll also offer some hypothetical problems for which hashing the inputs would be an appropriate solution:
- Security: You want to verify that the user has access to a secret, but you don't want to be able to read the secret directly.
- Data integrity: You want to verify that you have successfully obtained an exact copy of a large file.
- Mapping: You want to evenly distribute some collection of input data over an output space such that input data is essentially mapped to a random node, except that the node is consistent across multiple computations of the mapping. (For example implementing a dictionary, distributed database, or load balancer).

As far as I can tell, number 1 may actually be a disadvantage to your application, and number 2 and 3 are both attainable by simply writing a JSON file. So it is my opinion, that there is no reason to do any hashing here, and I would just write the inputs to a file using the code above.

**Comment by dbekaert** [comment]
@Askaholic for one of the variables that drive if there is a change is the cutline (json polygon) which depends on the following user input options: 
- bbox (SNWE, shapefile), this is user driven
- union/intersection, which depends on the footprints of the data being available and can this change of data is added.
- minimum overlap threshold (area in km^2)

There are different combination which could map to the same cutline. How would you recommend to handle this on to see if there is a change? The cutline gets computed as part based on these variables. The polygon itself might have machine precision changes that makes it almost identical.

**Comment by Askaholic** [comment]
I'd probably save the computed polygon along with the inputs, and just recompute it each time to compare against the saved version. You'll have to build in some sort of "isclose" functionality to check if the polygons are basically the same.

**Comment by dbekaert** [comment]
Ok that sounds pretty transparent, lets go down that route instead of the hash.

**Comment by mgovorcin** [comment]
closing, as it has been addressed in #454
